src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -1267,19 +1267,48 @@
   (let [f (first expr)]
     (and (str f) (= f "extract"))))
 
+(defn get-clause
+  [clause clauses]
+  (when-let [candidates (seq (filter #(= (first %) clause) clauses))]
+    (if (> (count candidates) 1)
+      (throw (IllegalArgumentException.
+               (format "Multiple %s clauses are not permitted" clause)))
+      (second (first candidates)))))
+
+(defn process-order-by
+  [clauses]
+  (when clauses
```

Comment:
validation question: should this be a valid query if there are no clauses? For me it silently works.